### PR TITLE
Various Cypher refactorings

### PR DESCRIPTION
**Pull request by SaschaPeukert**

### CODE REVIEW

File: community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/simplifyPredicates.scala [hunk 1]
```diff
@@ -22,11 +22,15 @@ package org.neo4j.cypher.internal.compiler.v3_3.planner.logical.plans.rewriter
...
 
-case object simplifyEquality extends Rewriter {
+case object simplifyPredicates extends Rewriter {
   override def apply(input: AnyRef) = instance.apply(input)
 
   private val instance: Rewriter = bottomUp(Rewriter.lift {
     case in@In(exp, ListLiteral(values@Seq(idValueExpr))) if values.size == 1 =>
       Equals(exp, idValueExpr)(in.position)
+
+    // This form is used to make finding composite index seeks and scans
```
Comment (fickludd) [inline community/cypher/cypher-compiler-3.3/src/main/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/simplifyPredicates.scala]: There is something strange in this sentence...

File: community/cypher/cypher-compiler-3.3/src/test/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/SimplifyPredicatesTest.scala [hunk 2]
```diff
@@ -42,6 +43,18 @@ class SimplifyEqualityTest extends CypherFunSuite with LogicalPlanningTestSuppor
     val orgPredicate: Expression = In(Property(varFor("x"), PropertyKeyName("prop")(pos))(pos), collection)(pos)
     val selection = Selection(Seq(orgPredicate), singleRow)(solved)
 
-    selection.endoRewrite(simplifyEquality) should equal(selection)
+    selection.endoRewrite(simplifyPredicates) should equal(selection)
+  }
+
+  test("should rewrite WHERE AndedPropertyInequality(x.prop, 1) to WHERE x.prop <> 1") {
+    val singleRow: LogicalPlan = Argument(Set(IdName("a")))(solved)(Map.empty)
+    val variable = Variable("a")(pos)
```
Comment (fickludd) [inline community/cypher/cypher-compiler-3.3/src/test/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/SimplifyPredicatesTest.scala]: I'm still so bad at reading these manually built logical plans... I guess `"a"` here really is what is called `x` in the test description?

File: community/cypher/cypher-compiler-3.3/src/test/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/SimplifyPredicatesTest.scala [hunk 3]
```diff
@@ -42,6 +43,18 @@ class SimplifyEqualityTest extends CypherFunSuite with LogicalPlanningTestSuppor
...
+    val property = Property(variable, PropertyKeyName("prop")(pos))(pos)
+    val greaterThan = GreaterThan(property, SignedDecimalIntegerLiteral("42")(pos))(pos)
```
Comment (fickludd) [inline community/cypher/cypher-compiler-3.3/src/test/scala/org/neo4j/cypher/internal/compiler/v3_3/planner/logical/plans/rewriter/SimplifyPredicatesTest.scala]: ... and `greaterThan` is `<>`, and `42` is `1`?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/ActualPipeBuilder.scala [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+import org.neo4j.cypher.internal.frontend.v3_3.helpers.Eagerly
+import org.neo4j.cypher.internal.frontend.v3_3.phases.Monitors
+import org.neo4j.cypher.internal.ir.v3_3.{IdName, VarPatternLength}
+import org.neo4j.graphdb.{Node, PropertyContainer, Relationship}
+
+/**
+ * Responsible for turning a logical plan with argument pipes into a new pipe.
+ * When adding new Pipes and LogicalPlans, this is where you should be looking.
+ */
+case class ActualPipeBuilder(monitors: Monitors, recurse: LogicalPlan => Pipe, readOnly: Boolean, idMap: Map[LogicalPlan, Id], expressionConverters: ExpressionConverters)
```
Comment (fickludd) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/ActualPipeBuilder.scala:47]: This class is a great idea. It summarizes the whole `Logical Plan -> Physical/Executable Plan` transformation very nicely.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+
+object PipelineInformation {
+  def empty = new PipelineInformation(Map.empty, 0, 0)
+
+  def apply(slots: Map[String, Slot], numberOfLongs: Int, numberOfReferences: Int) =
+    new PipelineInformation(slots, numberOfLongs, numberOfReferences)
+}
+
+class PipelineInformation(var slots: Map[String, Slot], var numberOfLongs: Int, var numberOfReferences: Int) {
```
Comment (fickludd) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/PipelineInformation.scala]: How about calling it `PipelineSlots`?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ */
+package org.neo4j.cypher.internal.compatibility.v3_3.runtime
+
+import org.neo4j.cypher.internal.compiler.v3_3.planner.logical.plans._
+import org.neo4j.cypher.internal.frontend.v3_3.InternalException
+import org.neo4j.cypher.internal.ir.v3_3.IdName
+
+import scala.collection.mutable
+
+object RegisterAllocation {
```
Comment (fickludd) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala]: And `SlotAllocation`? Then we get rid of register allocation overloaded meaning, and homogenize with `Slot`.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  def allocateRegisters(lp: LogicalPlan): Map[LogicalPlan, PipelineInformation] = {
```
Comment (fickludd) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala]: `allocateSlots` etc...

File: enterprise/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocationTest.scala [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    val expandAllocations = allocations(expand)
+    expandAllocations should equal(
+      PipelineInformation(Map("x" -> LongSlot(0), "r" -> LongSlot(1)), numberOfLongs = 2, numberOfReferences = 0))
+  }
+
+//  test("single node with two expands") {
+//    // given
+//    val root = AllNodesScan(IdName("x"), Set.empty)(solved)
+//    val e1 = Expand(root, IdName("x"), SemanticDirection.INCOMING, Seq.empty, IdName("z"), IdName("r1"), ExpandAll)(solved)
+//    val e2 = Expand(e1, IdName("z"), SemanticDirection.INCOMING, Seq.empty, IdName("y"), IdName("r2"), ExpandAll)(solved)
```
Comment (fickludd) [inline enterprise/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocationTest.scala]: Ghost code! Expel it with water, rinse and repeat.

**Review by fickludd — CHANGES_REQUESTED** [review]
I think this PR i a little bit weird, in that is contains some variable offset allocation code in addition to the "Various Cypher refactorings". Would it be possible to break out the offset allocation code into a separate PR? Possibly that part want's some more intergration tests too.

The refactorings all seem fine and reasonable!

I did make some comments on the offset code (and suggest that we settle for Slot as the gathering term), but don't know how things will be grouped in the end.

**Comment by SaschaPeukert** [comment]
I removed the two last register allocation commits because they were not meant to be included in this PR

**Review by fickludd — APPROVED** [review]

**Pull request merged**